mox/impl/src/lib.rs

```diff
@@ -1,245 +1,239 @@
+                Err(Error::new(punctuated.span(), "Dash tag name syntax isn't supported"))
+            }
+            NodeName::Colon(punctuated) => {
+                Err(Error::new(punctuated.span(), "Colon tag name syntax isn't supported"))
+            }
+        }
     }
 }
 
-impl ToTokens for MoxTag {
-    fn to_tokens(&self, tokens: &mut TokenStream) {
-        tag_to_tokens(&self.name, &self.attributes, Some(&self.children), tokens);
+fn mangle_expr_path(name: &mut syn::ExprPath) {
+    for segment in name.path.segments.iter_mut() {
+        mangle_ident(&mut segment.ident);
     }
 }
 
-fn tag_to_tokens(
-    name: &Ident,
-    attributes: &[MoxAttr],
-    children: Option<&[MoxItem]>,
-    stream: &mut TokenStream,
-) {
-    // this needs to be nested within other token groups, must be accumulated
-    // separately from stream
-    let mut contents = quote!();
+impl TryFrom<syn_rsx::Node> for MoxAttr {
+    type Error = syn::parse::Error;
 
-    for attr in attributes {
-        attr.to_tokens(&mut contents);
-    }
-
-    if let Some(items) = children {
-        for child in items {
-            MoxAttr::child(child).to_tokens(&mut contents);
+    fn try_from(node: syn_rsx::Node) -> syn::Result<Self> {
+        use syn_rsx::NodeType::*;
+        match node.node_type {
+            Element | Text | Block => Err(node_parse_error(&node, "MoxAttr")),
+            Attribute => {
+                Ok(MoxAttr { name: MoxAttr::validate_name(node.name.unwrap())?, value: node.value })
+            }
         }
     }
-
-    quote!(mox::topo::call(|| { #name() #contents .build() })).to_tokens(stream);
 }
 
-impl From<SnaxTag> for MoxTag {
-    fn from(SnaxTag { name, attributes, children }: SnaxTag) -> Self {
-        let attributes = attributes.into_iter().map(MoxAttr::from).collect();
-        Self {
-            // TODO get the span for the whole tag
-            span: name.span(),
-            name,
-            attributes,
-            children: children.into_iter().map(MoxItem::from).collect(),
+impl MoxAttr {
+    fn validate_name(name: syn_rsx::NodeName) -> syn::Result<syn::Ident> {
+        use syn::{parse::Error, punctuated::Pair, PathSegment};
+        use syn_rsx::NodeName;
+
+        let invalid_error = |span| Error::new(span, "Invalid name for an attribute");
+
+        match name {
+            NodeName::Path(syn::ExprPath {
+                attrs,
+                qself: None,
+                path: syn::Path { leading_colon: None, mut segments },
+            }) if attrs.is_empty() && segments.len() == 1 => {
+                let pair = segments.pop();
+                match pair {
+                    Some(Pair::End(PathSegment { mut ident, arguments }))
+                        if arguments.is_empty() =>
+                    {
+                        mangle_ident(&mut ident);
+                        Ok(ident)
+                    }
+                    _ => Err(invalid_error(segments.span())),
```

Comment:
Could you spell out the different ways this can fail? IIUC that might be:

* pair.is_none()
* pair.unwrap() != Pair::End
* !arguments.is_empty()

Reply:
I see the only way attribute name can be is the `Ident`. I could not come up with a good phrasing for each of these fails.

Reply:
Do you know what proc macro input could trigger them? I'm imagining something like (respectively)

* `mox!(<foo ::="bar">)` -- I'm not sure if syn_rsx will parse this but it's technically a valid rust path IIUC
* `mox!(<foo path::based::attr="bar")` -- it would be good to support this actually
* `mox!(<foo attr::<Baz, Quux>="bar">` -- `syn` supports paths that include generic arguments

We probably want different errors for each of these, is my thinking.

Reply:
`syn-rsx` uses a custom parser to generate `ExprPath`s, in order to support reserved keywords (like `type`) - which aren't supported by the default `ExprPath` implementation. The custom parser only supports `Ident`s separated by `::` currently, everything else will produce the error "expected punctuated node name"

Reply:
That's a fun trick and good to know! I guess that implies a question: should mox attempt to report precise errors for the whole `syn` interface or only for those parts we know `syn-rsx` to produce?

I'm inclined towards errors that handle the full range of possible `syn` values here, unless you think it's basically out of scope for `syn-rsx` to ever emit those.

Reply:
I'm definitely interested in improving error handling/emitting in `syn-rsx`, though I'm not exactly sure how to approach that or what exactly might be useful. So if you have ideas or there's something lacking, please let me know

Reply:
I guess in this particular case the question was whether `syn-rsx` could provide granular errors for `<foo ::="bar">` and `<foo attr::<Baz, Quux>="bar">`? Hmm, wondering whether it might make sense to introduce an `thiserror`-powered `syn_rsx::Error` and have it include error type information along with a `Span`, so error messages could be controlled completely from `syn-rsx` consumers?

Reply:
Opened https://github.com/stoically/syn-rsx/issues/12 to track that

Reply:
Cool! @ZetaNumbers could you just add a TODO referencing that issue and you can resolve this?

Reply:
```suggestion
                    // TODO improve error handling, see `https://github.com/stoically/syn-rsx/issues/12`
                    _ => Err(invalid_error(segments.span())),
```